### ISSUE TITLE
snapshot crasp

### ISSUE THREAD

**Issue by jjkke88** [may0324/DeepCompression-caffe#1]
when snapshot start, it crash.
```I0721 11:04:23.275279 17167 solver.cpp:228] Iteration 9, loss = 1.88784
I0721 11:04:23.275313 17167 solver.cpp:244]     Train net output #0: acc = 0.428571
I0721 11:04:23.275321 17167 solver.cpp:244]     Train net output #1: loss = 1.88784 (* 1 = 1.88784 loss)
I0721 11:04:23.275328 17167 sgd_solver.cpp:106] Iteration 9, lr = 0.001
I0721 11:04:23.275475 17167 solver.cpp:454] Snapshotting to binary proto file dataset/lenet_iter_10.caffemodel```

Process finished with exit code 139 (interrupted by signal 11: SIGSEGV)

**Comment by daimagou** [comment]
我也拉了博主的代码，跑了下，没有遇到你这个问题，不过跑出来后，发现caffemodel和以前一样的大，不知道是不是哪里操作的不对，你最后有弄出来没有？

**Comment by may0324** [comment]
@jjkke88 保存时发生crash应该跟代码无关，可以检查一下是不是磁盘空间不够等问题

**Comment by may0324** [comment]
@daimagou 直接用caffe跑出来的模型依然是原始大小，因为模型依然是.caffemodel类型，虽然大部分权值为0且共享，但每个权值依然以32float型存储，故后续只需将非零权值及其下标以及聚类中心存储下来即可，这部分可参考作者论文，写的很详细。

**Comment by pgadosey** [comment]
any suggestions on how to store the non zero weights and cluster center?

**Comment by guozhongluo** [comment]
@may0324 根据你们的信息，model大小没变，但是预测速度有没有加快？

**Comment by guozhongluo** [comment]
@pgadosey pruning , predict can be speed up?

**Comment by xiaohu2015** [comment]
为什么ComputeBlobMask()没有被调用？

**Comment by guozhongluo** [comment]
@xiaohu2015 你读取的是.h5模型，要读取caffemodel格式模型，才会调用ComputeBlobMask()

**Comment by xiaohu2015** [comment]
我说的是在LayerSetUp方法内要对这个函数进行调用吧

**Comment by may0324** [comment]
@xiaohu2015 deep compression 需要读入已训练好的模型做压缩，在LayerSetUp函数中，每层仅仅做了内存分配但未读入已训模型，因此此时的模型参数是随机初始化的，需要在CopyTrainedLayersFrom函数中调用ComputeBlobMask函数

**Comment by may0324** [comment]
@pgadosey i will show my code later for reference

**Comment by may0324** [comment]
@guozhongluo 实测在PC端提速效果不明显，因为大量权值为0 采用矩阵加速库影响不会很大，移动端有一定的提升

**Comment by suke27** [comment]
@may0324  本次只是做了权值减枝，剪完后 模型大小木有变化. 那运行时内存应该也不会发生变化，预测速度呢，有木有变化

**Comment by may0324** [comment]
@suke27 pc端由于采用了矩阵加速库，所以提速不明显，移动端稍有提升

**Comment by jiaqun123** [comment]
@may0324 @daimagou @guozhongluo 
 在读取caffemodel参数时（前提是make matcaffe 已成功，且成功读取其它模型参数）使用MATLAB读取剪枝聚类后的caffemodel，出现如下错误
![image](https://user-images.githubusercontent.com/19303915/32211309-e1d7fdb8-be4c-11e7-9e70-f9e5494630eb.png)

然后又用Python读取，也发生错误，程序与错误如下
![image](https://user-images.githubusercontent.com/19303915/32211535-ea33ba3c-be4d-11e7-808d-e13f438b3f2a.png)

![image](https://user-images.githubusercontent.com/19303915/32211352-1baa28d6-be4d-11e7-8387-2225bfe347f3.png)
希望能得到你们的帮助，非常感谢~

**Comment by jiaqun123** [comment]
@pgadosey @may0324 
hi，can you share the code that how to store the non zero weights and cluster center for reference?
thanks very much

**Comment by pgadosey** [comment]
@jiaqun123 unforyunately, i never got around it. I found this repository that implemenets the quantization part of this paper but for inference purposes you have to dequantize your weights in order to use them. https://github.com/yuanyuanli85/CaffeModelCompression . Kindly let me know if you are able to store them in another way

**Comment by jiaqun123** [comment]
@pgadosey   thanks,  I found this repository that implemenets the storage of sparse matrix. https://github.com/ZhouYuSong/caffe-pruned . You can learn from it.

**Comment by MrLinNing** [comment]
![image](https://user-images.githubusercontent.com/24288811/34987931-de9b4bae-faf7-11e7-92ec-59df537d5a07.png)
hi，I  run the python script，why the final accuracy is 0.0974？

**Comment by xingruan** [comment]
怎么查看压缩效果呢，，压完文件大小还是1.7M呀？

**Comment by xi-mao** [comment]
@may0324  我感觉这种方式只能减少在硬盘或者flash上的的存储大小啊，在实际运行时，0还是要恢复到内存中，而且即使加了一个mask，也需要一条判断语句，感觉效果应该不明显啊。有没有人能解释解释。

**Comment by xi-mao** [comment]
@jiaqun123   问一下 ，作者这个代码你是怎么编译运行的啊，还没用过caff，谢谢！
